benchexec/tools/ultimate.py

```diff
@@ -204,113 +186,120 @@ def _requires_ultimate_data(self, executable):
+            return self._cmdline_default(executable, combined_options, task)
 
-            cmdline += tasks
-            self.__assert_cmdline(
-                cmdline,
-                "cmdline contains empty or None argument when using SVCOMP17 mode: ",
-            )
-            return cmdline
-
-        if self._uses_propertyfile:
-            # use the old wrapper script if a property file is given
-            cmdline = [executable, "--spec", propertyfile]
-            if tasks:
-                cmdline += ["--file"] + tasks
-            cmdline += options
-            self.__assert_cmdline(
-                cmdline,
-                "cmdline contains empty or None argument when using default SVCOMP mode: ",
+        # there is no way to run ultimate; not enough parameters
+        msg = (
+            "Unsupported argument combination: You either need a property file or a toolchain option (-tc). "
+            "options={} property_file={} resource_limits={}".format(
+                options, task.property_file, resource_limits
             )
-            return cmdline
-
-        # if no property file is given and toolchain (-tc) is, use Ultimate directly
-        if "-tc" in options or "--toolchain" in options:
-            # ignore executable (old executable is just around for backwards compatibility)
-            mem_bytes = rlimits.get(MEMLIMIT, None)
-            cmdline = [self.java]
-
-            # -ea has to be given directly to java
-            if "-ea" in options:
-                options = [e for e in options if e != "-ea"]
-                cmdline += ["-ea"]
-
-            if mem_bytes:
-                cmdline += ["-Xmx" + str(mem_bytes)]
-            cmdline += ["-Xss4m"]
-            cmdline += ["-jar", self._get_current_launcher_jar(executable)]
-
-            if self._requires_ultimate_data(executable):
-                if "-ultimatedata" not in options and "-data" not in options:
-                    if self.api == 2:
-                        cmdline += [
-                            "-data",
-                            "@noDefault",
-                            "-ultimatedata",
-                            os.path.join(os.path.dirname(executable), "data"),
-                        ]
-                    if self.api == 1:
-                        raise ValueError(
-                            "Illegal option -ultimatedata for API {} and Ultimate version {}".format(
-                                self.api, self.version(executable)
-                            )
+        )
+        raise UnsupportedFeatureException(msg)
+
+    def _cmdline_no_wrapper(self, executable, options, task, resource_limits):
+        mem_bytes = resource_limits.memory
+        cmdline = [self.java]
+
+        # -ea has to be given directly to java
+        if "-ea" in options:
+            options = [e for e in options if e != "-ea"]
+            cmdline += ["-ea"]
+
+        if mem_bytes:
+            cmdline += ["-Xmx" + str(mem_bytes)]
```

Comment:
I noticed this code now and while the decision of how to do this is up to you, I wanted to mention the following (from my experience as a CPAchecker developer, not as BenchExec maintainer):

The memory limit from BenchExec (which is in `mem_bytes`) is a hard limit, reaching it will let the process be killed. Using it as the max heap size of a JVM will mean that the JVM might try to use that much memory for heap plus all the other memory (native allocations, stack, etc.), so it might well exceed the limit. Due to the GC, the JVM could even try to allocate that much memory even if it could cope with less memory (at the cost of more GC runs). So in CPAchecker, for example, we always set the max heap to some smaller value. How much smaller unfortunately depends on a lot of things, including how much memory is needed by non-JVM allocations (e.g., from a native SMT solver).

Reply:
Yes, we had the same policy with Ultimate for a long time. Nevertheless, it seems that allowing the VM in prinicpal to go up to this bound reduces OOMs, at least on SVCOMP instances.